src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -3146,9 +3146,6 @@ bool JvmtiSampledObjectAllocEventCollector::object_alloc_is_safe_to_sample() {
     return false;
   }
 
-  if (MultiArray_lock->owner() == thread) {
-    return false;
-  }
```

Comment:
So the recursive nature of the lock now means we don't have to bail out here - right?

Reply:
Yes.  If it were only the JNI upcall that was broken while holding the MultiArray_lock, I think I could have used this same approach.  Unfortunately, its also throwing OOM for metaspace and for Java heap :(